ios/sound.mm

```diff
@@ -24,209 +24,277 @@
-    midiInPortRef ( static_cast<MIDIPortRef> ( NULL ) )
+#define kOutputBus 0
+#define kInputBus  1
+
+/**
+ This callback is called when sound card needs output data to play. And because Jamulus use the same buffer to store input and output data (input is
+ sent to server, then output is fetched from server), we actually use the output callback to read inputdata first, process it, and then copy the
+ output fetched from server to ioData, which will then be played.
+ */
+static OSStatus recordingCallback ( void*                       inRefCon,
+                                    AudioUnitRenderActionFlags* ioActionFlags,
+                                    const AudioTimeStamp*       inTimeStamp,
+                                    UInt32                      inBusNumber,
+                                    UInt32                      inNumberFrames,
+                                    AudioBufferList*            ioData )
 {
-    NSError *audioSessionError = nil;
-  
-    [[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayAndRecord error:&audioSessionError];
-    [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted) {
-        if (granted) {
-            // ok
-        }
-    }];
-    [[AVAudioSession sharedInstance] setMode:AVAudioSessionModeMeasurement error:&audioSessionError];
-
-  //GetAvailableInOutDevices();
-}
 
-int CSound::Init ( const int /*iNewPrefMonoBufferSize*/ )
-{
+    CSound* pSound = static_cast<CSound*> ( inRefCon );
 
-  // store buffer size
-  //iCoreAudioBufferSizeMono = audioSession.IOBufferDuration;
+    AudioBuffer buffer;
 
-  // init base class
-  //CSoundBase::Init ( iCoreAudioBufferSizeMono );
+    buffer.mNumberChannels = 2;
+    buffer.mDataByteSize   = pSound->iCoreAudioBufferSizeMono * sizeof ( Float32 ) * buffer.mNumberChannels;
+    buffer.mData           = malloc ( buffer.mDataByteSize );
 
-  // set internal buffer size value and calculate stereo buffer size
-  //iCoreAudioBufferSizeStereo = 2 * iCoreAudioBufferSizeMono;
+    // Put buffer in a AudioBufferList
+    AudioBufferList bufferList;
+    bufferList.mNumberBuffers = 1;
+    bufferList.mBuffers[0]    = buffer;
 
-  // create memory for intermediate audio buffer
-  //vecsTmpAudioSndCrdStereo.Init ( iCoreAudioBufferSizeStereo );
+    // Obtain recorded samples
 
-  return iCoreAudioBufferSizeMono;
-}
+    // Calling Unit Render to store input data to bufferList
+    AudioUnitRender ( pSound->audioUnit, ioActionFlags, inTimeStamp, 1, inNumberFrames, &bufferList );
 
-void CSound::Start()
-{
-    // ??
+    // Now, we have the samples we just read sitting in buffers in bufferList
+    // Process the new data
+    pSound->processBufferList ( &bufferList, pSound ); // THIS IS WHERE vecsStereo is filled with data from bufferList
 
-    // call base class
-    //CSoundBase::Start();
-}
+    // release the malloc'ed data in the buffer we created earlier
+    free ( bufferList.mBuffers[0].mData );
```

Comment:
There should be _no_ (absolute) allocation and freeing of memory inside the audio processing loop.  It's a major source of performance hits.  Admittedly, I'm not familiar with how it's done for other platforms but if possible, allocate the buffer once when the app starts, then re-use it.

Reply:
You might be right. I was following this example: https://github.com/Epskampie/ios-coreaudio-example/blob/master/IosAudioController.m which really helped me.
About performance, I've been playing with my iPhone to listen to people and jam for hours and didn't really have any problem.
I'll keep that in mind and see if I can do anything to make it better.

Reply:
I was trying to use a pointer to store the buffer address, so that the buffer only needs to be initialized once and can be reused. However, since `recordingCallback` is a static function, the pointer must also be static. When compiling in xCode, I had this linker error (couldn't access that pointer from CSound or something).

Do you know how to overcome this error?

Reply:
I'd suggest making recordingCallback _not_ static, as a member of a single-instance class it should still be unique - so make sure the class it belongs to _is_ single-instance.  Then the buffer pointer should be a member of the same single-instance class, pointing to memory allocated when that class is instantiated.

I don't know iOS or ObjectiveC, though, to know how best to make that happen.

Reply:
I tried, but it seems the callback has to be static :(

Reply:
ha ha got it working! The secret is the pointer has to be called like this: `pSound->pointer`, not just `pointer`. Now 1 malloc only in constructor, which is freed in destructor. It seems to work, but I'm testing a bit before committing.

Reply:
Also I checked the MacOS sound code, and there's mallocs there ...